url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ShMemClient.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2013 Sergey Zubarev, info@js-labs.org
+ *
+ * This file is a part of JS-Collider framework.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.jsl.collider;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.BufferOverflowException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetEncoder;
+
+public class ShMemClient extends ShMem
+{
+    private final int m_blockSize;
+    private final ChannelIn m_in;
+    private final ChannelOut m_out;
+    private final ByteBuffer m_c2sBB;
+    private final ByteBuffer m_s2cBB;
+
+    public ShMemClient( String fileHint, int blockSize, File directory ) throws IOException
+    {
+        /* It will be better if block size is a multiplier of 4096. */
+        if ((blockSize & 0x0FFF) > 0)
+        {
+            blockSize &= ~0x0FFF;
+            blockSize +=  0x1000;
+        }
+        m_blockSize = blockSize;
+
+        final String prefix = "jsc-" + fileHint + "-";
+        final File fileC2S = File.createTempFile( prefix, ".c2s", directory );
+        m_out = new ChannelOut( fileC2S, blockSize, true );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when CharsetEncoder.encode throws an exception, the *ShMemClient* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *m\_out* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).